File: src/topic-03-divide-conquer/01_merge_sort/merge_sort.py
"""병합 정렬 — 반으로 나눠 각각 정렬하고, 정렬된 두 절반을 합친다.

실행: 편집기 오른쪽 위 ▶ 버튼, 또는 `python3 merge_sort.py`
"""


def merge_sort(a, lo=None, hi=None, temp=None, counts=None):
    """a[lo..hi]를 제자리에서 오름차순으로 정렬하고
    (비교 횟수, 이동 횟수)를 돌려준다."""
    if lo is None:
        lo, hi = 0, len(a) - 1
        temp = [0] * len(a)
        counts = [0, 0]  # [비교, 이동]
    if lo >= hi:
        return tuple(counts)
    mid = (lo + hi) // 2
    merge_sort(a, lo, mid, temp, counts)
    merge_sort(a, mid + 1, hi, temp, counts)
    _merge(a, lo, mid, hi, temp, counts)
    return tuple(counts)


def _merge(a, lo, mid, hi, temp, counts):
    """정렬된 a[lo..mid]와 a[mid+1..hi]를 하나로 합친다.

    temp로 한 번, 다시 a로 한 번 가므로 원소마다 이동 2회다.
    """
    i, j, k = lo, mid + 1, lo
    while i <= mid and j <= hi:
        counts[0] += 1
        if a[i] <= a[j]:  # <=라서 같은 값이면 왼쪽이 먼저: 안정
            temp[k] = a[i]
            i += 1
        else:
            temp[k] = a[j]
            j += 1
        counts[1] += 1
        k += 1
    while i <= mid:
        temp[k] = a[i]
        i += 1
        k += 1
        counts[1] += 1
    while j <= hi:
        temp[k] = a[j]
        j += 1
        k += 1
        counts[1] += 1
    for k in range(lo, hi + 1):
        a[k] = temp[k]
        counts[1] += 1


if __name__ == "__main__":
    a = [2, 8, 5, 9, 1, 10, 7, 6, 4, 3]

    print("before:", *a)
    comparisons, moves = merge_sort(a)
    print("after :", *a)
    print(f"comparisons = {comparisons}, moves = {moves}")

File: src/topic-03-divide-conquer/01_merge_sort/mergeSort.c
/* 병합 정렬 — 반으로 나눠 각각 정렬하고, 정렬된 두 절반을 합친다.
 *
 * 실행: 편집기 오른쪽 위 ▶ 버튼, 또는
 *   make src/topic-03-divide-conquer/01_merge_sort/mergeSort.out
 */
#include <stdio.h>

/* A[lo..mid]와 A[mid+1..hi]가 각각 정렬된 상태에서 하나로 합친다.
 * comparisons에는 비교 횟수를, moves에는 원소를 옮겨 적은 횟수를
 * 누적한다. temp로 한 번, 다시 a로 한 번 가므로 원소마다 이동 2회다. */
static void merge(int a[], int lo, int mid, int hi, int temp[],
                  int *comparisons, int *moves) {
    int i = lo;
    int j = mid + 1;
    int k = lo;
    while (i <= mid && j <= hi) {
        (*comparisons)++;
        if (a[i] <= a[j]) {   /* <=라서 같은 값이면 왼쪽이 먼저: 안정 */
            temp[k++] = a[i++];
        } else {
            temp[k++] = a[j++];
        }
        (*moves)++;
    }
    while (i <= mid) {
        temp[k++] = a[i++];
        (*moves)++;
    }
    while (j <= hi) {
        temp[k++] = a[j++];
        (*moves)++;
    }
    for (k = lo; k <= hi; k++) {
        a[k] = temp[k];
        (*moves)++;
    }
}

/* a[lo..hi]를 오름차순으로 정렬한다. */
void mergeSort(int a[], int lo, int hi, int temp[],
               int *comparisons, int *moves) {
    if (lo >= hi) {
        return;
    }
    int mid = lo + (hi - lo) / 2;
    mergeSort(a, lo, mid, temp, comparisons, moves);
    mergeSort(a, mid + 1, hi, temp, comparisons, moves);
    merge(a, lo, mid, hi, temp, comparisons, moves);
}

static void printArray(const char *label, const int a[], int n) {
    printf("%s:", label);
    for (int i = 0; i < n; i++) {
        printf(" %d", a[i]);
    }
    printf("\n");
}

int main(void) {
    int a[] = {2, 8, 5, 9, 1, 10, 7, 6, 4, 3};
    int temp[10];
    int n = sizeof(a) / sizeof(a[0]);
    int comparisons = 0;
    int moves = 0;

    printArray("before", a, n);
    mergeSort(a, 0, n - 1, temp, &comparisons, &moves);
    printArray("after ", a, n);
    printf("comparisons = %d, moves = %d\n", comparisons, moves);
    return 0;
}
